src/IROperator.cpp

```diff
@@ -20,7 +20,7 @@ namespace Halide {
 // terms, which is the degree plus one.
 namespace {
 
-Expr evaluate_polynomial(const Expr &x, float *coeff, int n) {
+Expr evaluate_polynomial(Expr x, float *coeff, int n) {
```

Comment:
I'm genuinely puzzled as to how this is an improvement. Doesn't this require making a copy to pass in? (Even though it's a cheap-ish refcount adjustment, that still seems more costly than a pass-by-ref.) Not trying to be argumentative, trying to understand.

Reply:
This function had to change one way or the other, because the std::move at the end of the body of it was a lie. The options were to remove the std::move, or change the signature so that the std::move actually did something useful. I changed the signature because there was one case where an rvalue was passed, and I figured we avoid a refcount bump in that case, and it's a wash in the lvalue case because we get charged a refcount bump on the call, but save one in the function body due to the std::move.

That's supposed to be the new general rule - pass by value and use std::move inside the function, but I find this hard to think about, so I'll just check the disassembly to see what happens...

Yeah, the body of evaluate_polynomial has two fewer refcount adjustments (on different if paths, so only one triggers), and the call-sites in the lvalue cases have an additional ref count bump, so it's a wash. The call-site in the rvalue case is unchanged, because the temporary x*x just becomes the argument directly without having to make a copy of it, so it's a savings in that case.

There's a little more complexity to it though, involving destructors:

In the pass-by-value case, the extra refcount bump at the lvalue call-sites is due to creating a temporary for the value argument, which then needs to be destroyed after the call (even though it has been moved from inside evaluate_polynomial and is actually just a default-constructed Expr() at that point), so there's also conditional refcount subtract and call to delete if it's zero. The condition is always false, so there's no actual atomic subtract or call to delete, but still, there's an extra branch, and a little extra code size, and the optimizer doesn't seem to realize it's dead code (to do that, it would have to look inside evaluate_polynomial and realize that the arg is unconditionally moved from, and that means the pointer is now null). In the pass-by-const-ref case the extra refcount bump is just because the returned value of evaluate_polynomial captures the argument one additional time, so it's just an add, there's no extra branch.

So the value case does fewer copies total, by moving some out of evaluate_polynomial to the call site instead, but the const-ref case does fewer destructions of moved-from-Exprs(), so it saves some branches and some code size. For types where a copy is expensive and destruction of a moved-from instance is cheap, accepting args by value is a big win if you ever pass an rvalue. Is Expr one of these types? Well, an atomic add is probably much more expensive that a perfectly-predicted branch, so I think it's a good trade-off.

I did actually experiment a while ago with making all of IR.h and IROperator.h take Exprs by const ref instead of by value and then not internally std::move, and it did indeed slow down the compiler due to the extra ref count bumps. Passing Exprs by value is better (slightly) when it's possible they could be rvalues, as they often are when writing expressions out.

Reply:
Shorter version: 

- A copy is inevitable because the Expr gets captured. 
- Passing by value instead of const-ref does one more copy at the call-site to create the Expr arg and one fewer copy inside the function (because now you can use std::move).
- It's better to have the copy at the call-site, because it can be optimized away in the case where the arg is a temporary.
- It does, however, cost additional code size at the call site.

Reply:
Thanks for the detailed investigation. LGTM.

Reply:
That was also the reasoning behind all the std::moves in the previous PR that changed Constructors to accept things by value. That was less worthwhile than having IROperator accept Exprs by value though, because those were almost never rvalues. It's very common for the functions in IROperator.